db/c.cc

```diff
@@ -341,6 +351,196 @@ struct rocksdb_compactionfilter_t : public CompactionFilter {
   bool IgnoreSnapshots() const override { return ignore_snapshots_; }
 };
 
+/* Table Properties */
+
+struct rocksdb_table_properties_t {
+  const TableProperties* rep;
+};
+
+struct rocksdb_table_properties_collector_factory_context_t {
+  TablePropertiesCollectorFactory::Context* rep;
+};
+
+struct rocksdb_user_collected_properties_t {
+  const UserCollectedProperties* rep;
+};
+
+struct rocksdb_user_collected_properties_mut_t {
+  UserCollectedProperties* rep;
+};
+
+struct rocksdb_table_properties_collector_t : public TablePropertiesCollector {
+  void* state_;
+  void (*destructor_)(void*);
+  // todo(pavel): support returning full Status
```

Comment:
This will probably take the form of passing in an extra pointers to store an optional message, but I really didn't want to go down the rabbit hole just yet.